### ISSUE TITLE
Clock loses 1.3-3.5 seconds per hour

### ISSUE THREAD

**Issue by nico** [SerenityOS/serenity#3429]
Now that we can set time via `ntpquery -s` fairly accurately (it's a one-shot adjustment), we can evaluate how well serenity's clock keeps time.

I ran `su; ntpquery -s; for i in $(seq 1000) { sleep 1; echo $i; ntpquery }` to see how much the clock gains over 1000 seconds, and it gained 3.5s. Then I tried `su; ntpquery -s; for i in $(seq 12) { sleep 1800; echo $i; ntpquery }` to check every 30 min over 6 hours. At the end of the 6h, the clock was 13s late.

After 20h of uptime, I checked again, by then it was 32.8s late.

Eventually we'll have a long-running NTP service that keeps adjusting the clock, but maybe the OS should be able to keep time better even without an internet connection.

--

I read the timekeeping code a bit, and I think we use the HPETComparator code, which should tick at 1000 Hz even. TimeManagement::epoch_time() assumes 1000Hz exactly instead of using `ticks_per_second()` but at least for the HPET that's not a source of error. So not sure where exactly we lose time, but I figured I'd file at least a bug for it.

**Comment by nico** [comment]
Oh, I should paste the raw data from my 6h experiment described above:

<details>

```
1
Offset: 1.256666
2
Offset: 2.393099
3
Offset: 3.388525
4
Offset: 4.455085
5
Offset: 5.387008
6
Offset: 6.299331
7
Offset: 7.236079
8
Offset: 8.239125
9
Offset: 9.368290
10
Offset: 10.399584
11
Offset: 11.694160
12
Offset: 13.049532

# later...
uptime
Up 20 hours, 3 minutes, 6 seconds
Offset: 32.787009
```
</details>

**Comment by nico** [comment]
Now at `20 hours, 56 minutes, 11 seconds` uptime, at an offset of 42.54s -- so lost almost 10s in less than an hour. The system probably loses more time when it's getting used -- maybe we drop interrupts or something like that.

**Comment by tomuta** [comment]
I don't think the "1000 Hz even" will be exactly that. All clock chips are subject to variation, differences due to manufacturing, quality, and temperature. Isn't that precisely the point of NTP so that the kernel can be periodically supplied with frequency adjustments to keep it roughly at the same rate as actual wall time?

**Comment by nico** [comment]
Sure, but 1-10s per hour seems still kind of bad maybe? That's worse than even a mechanical watch.

**Comment by tomuta** [comment]
I'm not sure how much is considered bad. I guess 30s of 20h is kind of bad...

**Comment by awesomekling** [comment]
Yeah that's bad. Would be good to get to the bottom of what's going on.
(And it would be good to see if other systems can keep time while running in QEMU) :)

**Comment by nico** [comment]
To see if cpu load has a big effect here, I ran `yes` in one terminal to use 100% cpu, and then did the same `ntpquery -s`, wait an hour, `ntpquery` dance again. This time, the system lost 7.3s in an hour -- that's more, but not super dramatically more.

**Comment by nico** [comment]
Today, after 18h21m uptime having System Monitor and Profiler open the entire time, the clock is 417s (ie almost 7 minutes) late, so almost 22s lost per hour.

**Comment by tomuta** [comment]
> Today, after 18h21m uptime having System Monitor and Profiler open the entire time, the clock is 417s (ie almost 7 minutes) late, so almost 22s lost per hour.

Yeah that's pretty bad...

**Comment by nico** [comment]
This gets worse if the _host_ is under high load, eg by runing `stress -c $(nproc)`.

This makes some sense: Time keeping relies on timer interrupts, but in qemu these are emulated and need qemu to be scheduled. If the system is under load, qemu will be scheduled less, and interrupts will fire less.

One possible improvement is to use a timekeeping method that doesn't require interrupts, e.g. rdtscp.

This theory suggests this is more a problem when running in an emulator than on actual hardware -- but that's how most people currently use serenityos.

**Comment by tomuta** [comment]
So after having worked on the HPET fixes in #3954 I have a few more theories and suggestions:

1. It seems that we fire timers 1000/1280 times per seconds. That seems overly aggressive and I think a whole lot of unnecessary overhead. I believe Linux' `jiffies` by default increment ~ 250 per second (though the kernel can support other frequencies, including 1000, for special use cases. It seems like we should change `m_system_timer` (the one used for scheduling purposes) probably should fire ~ 250 per second
2. Right now, for time keeping purposes, we increment the uptime on each timer irq. I wonder if due to 1) we may miss some interrupts, which could also be caused by a prolonged time of disabled interrupts (e.g. spinlocks etc). I suggest we also reduce that frequency dramatically, or disable them altogether (in the event that we use APIC timers) and instead base the time off of HPET's main counter, which can be queried at any given time. HPET timer frequency >= 1ms is supposed to be 0.05% accurate. So, essentially a `clock_gettime` would simply query the HPET's main counter, and that timer would not fire any IRQs in the case of SMP (which would just use the APIC timer).

Thoughts?

**Comment by nico** [comment]
Various timer-related things:

- As far as I understand, the timer interrupt sets the minimum scheduling quantum. For this, 1000 Hz seems high -- but if you're a game that tries to complete a game tick every 16ms, only being scheduled at a granularity of 4ms is probably challenging too.

- The system tick rate also determines the timer resolution for measurements. Being able to measure time with a precision of 1 ms seems good.

- IIRC most docs I've read recommend using HPET / ACPI to calibrate the TSC and then use RDTSCP for timekeeping (at least if CONSTANT_TSC / NONSTOP_TSC are available). Some care must be taken to detect deep sleep here. This would completely solve the "missed interrupt" problem, which according to https://en.wikipedia.org/wiki/High_Precision_Event_Timer#Problems can be a problem with HPET too.

- "base time off hpet" is a bit hand-wavy :) It must allow for things like time adjustments from NTP (I just uploaded some early primitive work in that area here: https://github.com/SerenityOS/serenity/pull/4025)

But I don't much real-world experience, this is just from reading around a bit.

**Comment by tomuta** [comment]
I think basing time off hpet would still allow you to adjust for NTP. But instead of applying such adjustment 1000/1280 times per second, you'd just interpolate (linear regression) based on the last time the interrupt fired. disclaimer: clock synchronization over networks is a part of my day job :-)

**Comment by tomuta** [comment]
As for TSC, I'm not sure rdtscp is a good/reliable way. Especially on more modern processors that may constantly drop/increase the clock frequency as needed, to save power. Based on my searching in that area recently (trying to calibrate the apic timer interrupts) it seemed that the conclusion is to not use it for the purpose of measuring time.

**Comment by nico** [comment]
If the processor has CONSTANT_TSC, rdtsc runs at a constant rate independent of turbo state. See links in commit message of 8a01be48491383d80213bdcb7c5b4e3d3c3ec624 for some pointers.

**Comment by tomuta** [comment]
With #4320 being merged there is a chance it works better now because lost interrupts should not cause any time drifts anymore. Time should be as accurate as HPET can be now.

**Comment by tomuta** [comment]
Is this still an issue? Can we close this?
